### ISSUE TITLE
Question about "Metric Learning Reality Check" paper

### ISSUE THREAD

**Issue by jac99** [KevinMusgrave/powerful-benchmarker#23]
Hi,

I'm not sure if this is an appropriate place, but I'd like to ask a question about your "A metric learning reality check" paper. It's an excellent work with great practical applications. Seems that a lot of 'progress' in deep metric learning field is due to some tweaking of parameters or flawed training/evaluation protocol.

But after reading the paper one thing is not clear to me. How training entities (pairs or triplets) are constructed? What, if any, miner (hard negative mining) is used? Specifically how pairs used in Contrastive loss and triplets in Triplet loss are constructed in each batch (containing C * M elements)? Are all possible pairs or triplets constructed for each batch? Or there's some form of hard negative mining used?

Maybe reported improvements in recent deep learning methods were due to better mining schemes, not to the formulation of the loss function itself?

**Comment by KevinMusgrave** [comment]
Thanks for your interest! 

I focused on loss functions to limit the scope of the experiments.

- For all tuple-based methods, all pairs/triplets in a batch are utilized. 
- The exception is MS+Miner, which uses the mining method proposed in the multi-similarity paper. 

In the original papers:

- With the exception of the multi-similarity loss and margin loss, I think the methods I tested did not use mining.
- The Margin loss paper indicates that even without mining, their loss is better than the contrastive and triplet loss. So I tested the margin loss without their distance weighted sampling. Based on the results of MS vs MS+Miner, my guess is that distance weighted sampling will make no difference. But even if it did, you might as well apply it to a different loss function because the adaptive margin in the Margin loss doesn't provide any advantage.
